### ISSUE TITLE
multiple monitors group switching

### ISSUE THREAD

**Issue by diskroma** [infokiller/i3-workspace-groups#29]
Currently, the behaviour for multiple monitors and groups is, that when switching to another group, the inactive monitor does not switch to the new group.
This is confusing in the workflow and you always need to swap workspaces for each screen.
Is this a bug, or intented behaviour?

I think it would be better if, by switching groups, all monitors switch to the same group so you have coherent behaviour.

**Comment by infokiller** [comment]
How do you switch to another group?
By default, if you don't use the `--focused-monitor-only` argument, the group is switched in every monitor **if it exists in that monitor**.

**Comment by diskroma** [comment]
Youre right, it switches the group, but it doesnt automatically open a workspace on that group.
I believe it would be more effective to switch to an active workspace in that group on all monitors, not just the one, from which the group switch was triggered.
Is that possible ?

Also, in that regard wouldnt it be nice to automatically open a new workspace, if theres none on the inactive monitor yet ?

**Comment by infokiller** [comment]
The group switching should apply to the focused monitor and all monitors that have that group. If a monitor doesn't have a group and is not focused, the group switch will not affect it. I think this should stay the default behavior.
I can add an option to also open a new workspace for the group if it doesn't exist in a monitor, but it may be a bit tricky, because I think the only way to create a new workspace is by focusing on it, so it will cause a focus switch.
Anyway, let's keep this open to see if there are other people that may be interested in that.

**Comment by Syphdias** [comment]
Hi, to be honest the fact that the groups are not global confuses me a bit.

Is there an option to ignore outputs? Like only one selected/active group no matter what output it is on.

**Comment by infokiller** [comment]
I'm not sure I understand the confusion, can you please clarify what confuses you and how you would want it to behave?

**Comment by Syphdias** [comment]
Hm, I think I caused that confusion myself.
* I set my `$ws10` to `10:Audio`
* I assign this workspace to my secondary output
* I assigned spotify to open on this output
* I try to switch to `10`/`Audio` but it opens a new workspace `10` (or rather `100010`)

With workspace numbers on it's more obvious what happened. I guess this is due to the missing `:10` suffix.

I expected this behavior to also cause problems when moving workspaces to other outputs, which I do on a regular basis. But I couldn't to reproduce this behavior by moving workspaces from one output to the other (thus "corrupting" the workspace numbers). This does cause sorting issues though. For example (with workspace numbers):
* `4:4`
* `5:bla:5`
* `100001:1`
When it should be:
* `100001:1`
* `100004:4`
* `100005:bla:5`

**Comment by infokiller** [comment]
> Hm, I think I caused that confusion myself.
> 
> * I set my `$ws10` to `10:Audio`

I think using a workspace number and name like this can cause issues, because as you already noticed, i3-workspace-groups needs to use the workspace numbers in order to control the order of groups and the workspace within groups.
See also #31

> * I assign this workspace to my secondary output
> * I assigned spotify to open on this output
> * I try to switch to `10`/`Audio` but it opens a new workspace `10` (or rather `100010`)

How do you do this switch?

> 
> With workspace numbers on it's more obvious what happened. I guess this is due to the missing `:10` suffix.

That suffix is the "local workspace number", which is the number of the workspace within the group (so that you can have workspace with the number 10 in multiple groups).

> 
> I expected this behavior to also cause problems when moving workspaces to other outputs, which I do on a regular basis. But I couldn't to reproduce this behavior by moving workspaces from one output to the other (thus "corrupting" the workspace numbers). This does cause sorting issues though. For example (with workspace numbers):

Yes, not that I think about it, moving workspaces between outputs using the native i3 commands will cause issues, because i3-workspace-groups needs to be aware of it.
I assume you're using `move workspace to output <output>` to move the workspaces?

> 
> * `4:4`
> * `5:bla:5`
> * `100001:1`
>   When it should be:
> * `100001:1`
> * `100004:4`
> * `100005:bla:5`
> 
> Edit: `i3-autoname-workspaces` can fix the sorting apparently.

**Comment by Syphdias** [comment]
> > * I assign this workspace to my secondary output
> > * I assigned spotify to open on this output
```i3
set $ws0 "10:Audio"
workspace $ws0 output DP-0
for_window [class="Spotify"] move to workspace $ws0
```
> > * I try to switch to `10`/`Audio` but it opens a new workspace `10` (or rather `100010`)
> 
> How do you do this switch?
```i3
set $exec_i3_groups exec --no-startup-id ~/.local/bin/i3-workspace-groups
bindsym $mod+1 $exec_i3_groups workspace-number 1
```

> I assume you're using `move workspace to output <output>` to move the workspaces?

Yes, I bound `left` and `right`.

> Yes, now that I think about it, moving workspaces between outputs using the native i3 commands will cause issues, because i3-workspace-groups needs to be aware of it.

It works fine and if you keep the renaming daemon running it works. By the way, the README doesn't even mention it :wink:

**Comment by infokiller** [comment]
@Syphdias just curious, are you still using i3-workspace-groups, or were the issues you ran into too annoying?

**Comment by Syphdias** [comment]
Yes, I switched to the keybindings but only at one of my computers. I'm still trying to use assignments unsuccessfully by trying to set `$ws0` to `10:Audio:10` to trick i3-workspace-groups to think it is on the primary monitor, in the default group, named Audio with id 10. Sadly this behaves nothing like I would expect. :sweat_smile:

If I rename a workspace with `i3-rename-workspace` it gets the same name but this time around it works just fine like I would expect. I wonder if there is some confusion due to the naming being similar to: group "Audio", id "10"

Some background information why I mainly use assignments. I try to group all media related programs to the same workspace and have a workspace dedicated for assigning games to – this is needed since lots of games don't place nicely with tiling windows and force their fullscreen without the possibility of getting to the windows on that workspace. That's why I get the class of new games with `xprop` and assign them to `11:Games` or `11:Games:11` (But like with Audio `i3-workspace-groups workspace-number 11` does not move me there)

**Comment by infokiller** [comment]
Do you open the media/games once in their dedicated workspace and then just leave them there and forget about them?

One workaround for having a program assigned to a specific workspace is to first open that workspace (for example, by switching to the `Games` group and then optionally rename it and change its number), and only then open the program. I would think this workaround is reasonably if you don't open these programs frequently, but it's annoying if you do. Am I right?
